gtda/images/filtrations.py

```diff
@@ -1036,3 +1038,227 @@ def plot(Xt, sample=0, colorscale='greys', origin='upper',
             title=f"Signed-distance filtration of image {sample}",
             plotly_params=plotly_params
             )
+
+
+@adapt_fit_transform_docs
+class DensityFiltration(BaseEstimator, TransformerMixin, PlotterMixin):
+    """Filtrations of 2D/3D binary images based on the number of neighboring
+    activated pixels.
+
+    The density filtration assigns to each pixel of a binary image a greyscale
+    value equal to the number of activated pixels with a ball centered around
+    it.
+
+    Parameters
+    ----------
+    radius : float, default: ``1``
+        The radius of the ball within which the number of activated pixels is
+        considered.
+
+    metric : string, or callable, optional, default: ``'euclidean'``
+        Determines a rule with which to calculate distances between
+        pairs of pixels.
+        If ``metric`` is a string, it must be one of the options allowed by
+        ``scipy.spatial.distance.pdist`` for its metric parameter, or a metric
+        listed in ``sklearn.pairwise.PAIRWISE_DISTANCE_FUNCTIONS``, including
+        "euclidean", "manhattan", or "cosine".
+        If ``metric`` is a callable function, it is called on each pair of
+        instances and the resulting value recorded. The callable should take
+        two arrays from the entry in `X` as input, and return a value
+        indicating the distance between them.
+
+    metric_params : dict, optional, default: ``{}``
```

Comment:
Don't we avoid using mutable defaults elsewhere? E.g. in amplitude where there is also a `metric_params`. Here, we would use `None` I guess.

Reply:
This is only if we are to modify it in `fit`, no? Here, we merely pass it to the right `sklearn` function.

Reply:
I'm not sure. To be honest, this seems to be a general pattern in `scikit-learn` (see https://www.google.com/search?client=firefox-b-d&q=scikit-learn+metric_params). I'd rather we stick if that's the case even if it's not necessary.

Reply:
`NeighborsBase`'s `fit` method modifies the `metric_params` so that's why all KNN-based estimators have this, but in general, if we don't modify them at all, it is not required. For some reason, ``None`` is also the default `metric_params` of `gtda.images.RadialFiltration`, so I could do it just out of consistency, or I could fix `RadialFiltration`.

The point is that we are generating an attribute `effective_metric_params` that is always equal to the hyperparameter `metric_params`.

Reply:
Thanks for the clarification. Indeed, in scikit-learn `metric_params` seems to be only used when `NeighborsBase` needs to be involved, I had not known that!

Indeed it seems unnecessarily cumbersome to have the `effective_metric_params_` attribute in `RadialFiltration` as it brings nothing to the table. I'm happy for you to fix this in `RadialFiltration` even if in an ideal world one would have perfect consistency.